File: gofft/alg/src/dsp.c
#include <stdlib.h>
#include "dsp.h"

double goertzel(double* data, long data_len, int fs, double ft, 
                int filter_size)
{
    double k;		// Related to frequency bins
    double omega;
    double sine, cosine, coeff, sf, mag;
    double q0, q1, q2, real, imag;
    long int i;
    long int dlen;

    k = floor(0.5 + ((double)(filter_size*ft) / (double)fs));

    omega = 2.0*M_PI*k/(double)filter_size;
    sine = sin(omega);
    cosine = cos(omega);
    coeff = 2.0*cosine;
    sf = (double)data_len;		// scale factor: for normalization

    q0 = 0.0;
    q1 = 0.0;
    q2 = 0.0;

    dlen = data_len - data_len%3;
    for (i = 0; i < dlen; i+=3)
    {
        q0 = coeff*q1 - q2 + data[i];
        q2 = coeff*q0 - q1 + data[i+1];
        q1 = coeff*q2 - q0 + data[i+2];
    }
    for (; i < data_len; i++)
    {
        q0 = coeff*q1 - q2 + data[i];
        q2 = q1;
        q1 = q0;
    }

    real = (q1 - q2*cosine)/sf;
    imag = (q2*sine)/sf;
    mag = sqrt(real*real + imag*imag);

    return mag;
}

void goertzel_m(double* data, long int data_len, int fs, double* ft, 
                int ft_num, int filter_size, double* mag)
{
    double k;
    double omega;
    double sine, cosine, coeff, sf;
    double q0, q1, q2, real, imag;
    long int i, dlen;
    int cnt;

    for (cnt = 0; cnt < ft_num; cnt++)
    {
        k = floor(0.5 + ((double)(filter_size*ft[cnt]) / (double)fs));
        omega = 2.0*M_PI*k/(double)filter_size;
        sine = sin(omega);
        cosine = cos(omega);
        coeff = 2.0*cosine;
        sf = (double)data_len;

        q0 = 0.0;
        q1 = 0.0;
        q2 = 0.0;

        dlen = data_len - data_len%3;
        for (i = 0; i < dlen; i+=3)
        {
            q0 = coeff*q1 - q2 + data[i];
            q2 = coeff*q0 - q1 + data[i+1];
            q1 = coeff*q2 - q0 + data[i+2];
        }
        for (; i < data_len; i++)
        {
            q0 = coeff*q1 - q2 + data[i];
            q2 = q1;
            q1 = q0;
        }

        real = (q1 - q2*cosine)/sf;
        imag = (q2*sine)/sf;
        mag[cnt] = sqrt(real*real + imag*imag);
    }
}

double goertzel_rng(double* data, long data_len, int fs, double ft, 
                    int filter_size, double rng)
{
    double f_step, f_step_normalized, k_s, k_e;
    double omega, sine, cosine, coeff, mag;
    double real, imag;
    double sf;
    double q0, q1, q2;
    double k;
    double f;
    long int i;
    long int dlen;

    f_step = (double)fs/(double)filter_size;
    f_step_normalized = 1.0/(double)filter_size;
    k_s = floor(0.5+ft/f_step);
    k_e = floor(0.5+(ft+rng)/f_step);
    sf = (double)data_len;

    mag = 0.0;

    for (k=k_s; k<k_e; k+=1.0)
    {
        f = k*f_step_normalized;
        omega = 2.0*M_PI*f;
        sine = sin(omega);
        cosine = cos(omega);
        coeff = 2.0*cosine;

        q0 = 0.0;
        q1 = 0.0;
        q2 = 0.0;

        dlen = data_len - data_len%3;
        for (i = 0; i < dlen; i+=3)
        {
            q0 = coeff*q1 - q2 + data[i];
            q2 = coeff*q0 - q1 + data[i+1];
            q1 = coeff*q2 - q0 + data[i+2];
        }
        for (; i < data_len; i++)
        {
            q0 = coeff*q1 - q2 + data[i];
            q2 = q1;
            q1 = q0;
        }

        real = (q1 - q2*cosine)/sf;
        imag = (q2*sine)/sf;
        mag += sqrt(real*real + imag*imag);

    }
    return mag;
}

File: gofft/alg/dsp.py
import numpy as np
from scipy.fftpack import fft as scipyfft
import dsp_ext as cext


__all__ = ['goertzel', 'goertzel_m', 'goertzel_st', 
           'goertzel_st_m', 'fft_eval', 'stfft_eval']

def goertzel(data, fs, ft, width, rng=None):
    """
    Goertzel algorithm, an efficiency method to evaluate specific terms of a
    discrecte Fourier transform.

    Parameters
    ----------
    data : ndarray
        Input signal.
    fs : int
        Sampling frequency.
    ft : int
        Target frequency.
    width : int
        Width of filter. (related to frequency resolution)
    rng : ndarray
        Frequency range for evaluation.

    Returns
    -------
    mag : float
        Magnitude of a single DFT term corresponding to target frequency.

    Note
    ----
    1. To get a better result, target frequency should not be too low.
    2. The higher `N` is, the better freqency resolution we can get, however, 
       with longer computation time will be taken.
    ref : https://en.wikipedia.org/wiki/Goertzel_algorithm
    """
    if fs > len(data):
        raise ValueError(
            'Data length is too short:{0}'.format(len(data)))

    if width > len(data):
        raise ValueError(
            'Size of Goertzel block(N) should be less than data length.')

    if data.dtype != np.dtype('float'):
        data = data.astype('float')
    ft = np.asfarray(ft)

    try:
        if rng:
            val = cext.goertzel_rng(data, fs, ft, width, rng)
        else:
            val = cext.goertzel(data, fs, ft, width)
    except:
        raise

    return val


def goertzel_m(data, fs, ft, width):
    """
    Modified Goertzel algorithm. This method evaluate all `ft` at once.

    Parameters
    ----------
    data : ndarray
        Input signal.
    fs : int
        Sampling frequency.
    ft : ndarray
        Target frequency.
    width : int
        Width of filter. (related to frequency resolution)

    Returns
    -------
    mag : ndarray
        Magnitude of a single DFT term corresponding to target frequency.
    """
    if fs > len(data):
        raise ValueError(
            'Data length is too short:{0}'.format(len(data)))

    if width > len(data):
        raise ValueError(
            'Size of Goertzel block(N) should be less than data length.')

    if data.dtype != np.dtype('float'):
        data = np.asarray(data, dtype='float')
    ft = np.asfarray(ft)

    try:
        val = cext.goertzel_m(data, fs, ft, width)
    except:
        raise

    return val


def goertzel_st(data, fs, ft, width, rng=None, padding=False):
    """
    Short-time Goertzel algorithm.

    Parameters
    ----------
    data : ndarray
        Input signal.
    fs : int
        Sampling frequency.
    ft : ndarray
        Target frequency.
    width : int
        Width of filter. (related to frequency resolution)
    rng : ndarray
        Frequency range for evaluation.
    padding : bool
        Apply padding for this algorithm.

    Returns
    -------
    val : ndarray
        Magnitude of a single DFT term corresponding to target frequency.
    """
    if fs > len(data):
        raise ValueError(
            'Data length is too short:{0}'.format(len(data)))

    if width > len(data):
        raise ValueError(
            'Size of Goertzel block(N) should be less than data length.')

    if data.dtype != np.dtype('float'):
        data = np.asarray(data, dtype='float')
    ft = np.asfarray(ft)

    rem = len(data)%width
    dlen = len(data)-rem
    val = 0.0
    cnt = 0
    for i in range(0, dlen, width):
        cnt += 1
        val += cext.goertzel(data[i:i+width], fs, ft, width)

    if rem!=0 and padding:
        cnt += 1
        pdata = np.zeros(width-rem, dtype='float')
        pdata = np.append(data[i+width:], pdata)
        val += cext.goertzel(pdata, fs, ft, width)

    val /= cnt
    return val


def goertzel_st_m(data, fs, ft, width, padding=False):
    """
    Modified short-time Goertzel algorithm. This method evaluates all `ft` 
    at once.

    Parameters
    ----------
    data : ndarray
        Input signal.
    fs : int
        Sampling frequency.
    ft : ndarray
        Target frequency.
    width : int
        Width of filter. (related to frequency resolution)
    padding : bool
        Apply padding for this algorithm.

    Returns
    -------
    val : ndarray
        Magnitude of a single DFT term corresponding to target frequency.
    """
    if fs > len(data):
        raise ValueError(
            'Data length is too short:{0}'.format(len(data)))

    if width > len(data):
        raise ValueError(
            'Size of Goertzel block(N) should be less than data length.')

    if data.dtype != np.dtype('float'):
        data = np.asarray(data, dtype='float')
    ft = np.asfarray(ft)

    rem = len(data)%width
    dlen = len(data)-rem
    val = 0.0
    cnt = 0
    for i in range(0, dlen, width):
        cnt += 1
        val += cext.goertzel_m(data[i:i+width], fs, ft, width)

    if rem!=0 and padding:
        cnt += 1
        pdata = np.zeros(width-rem, dtype='float')
        pdata = np.append(data[i+width:], pdata)
        val += cext.goertzel_m(pdata, fs, ft, width)

    val /= cnt
    return val


def fft_eval(sig, fs, ft):
    """
    Evaluate DFT terms of given tagert frequency.

    Parameters
    ----------
    sig : ndarray
        Input signal.
    fs : int
        Sampling frequency.
    ft : int, float or array-like
        Target frequency to be evaluated.

    Returns
    -------
    mag : ndarray
        Evaluated DFT terms.
    """
    ft = np.asfarray(ft)
    dlen = sig.size
    spec = scipyfft(sig) / dlen
    idx = (ft/fs*dlen).astype('int')
    mag = np.abs(spec[idx])
    return mag


def stfft_eval(sig, fs, ft, width):
    """
    Short-time version of `fft_eval()`.

    Parameters
    ----------
    sig : ndarray
        Input signal.
    fs : int
        Sampling frequency.
    ft : int, float or array-like
        Target frequency to be evaluated.
    width : int
        Window size for short-time technique.

    Returns
    -------
    mag : ndarray
        Evaluated DFT terms.
    """
    ft = np.asfarray(ft)
    rem = sig.size % width
    dlen = sig.size - rem
    cnt = sig.size // width
    mag = 0.0

    for i in range(0, dlen, width):
        spec = scipyfft(sig[i:i+width]) / width
        idx = (ft/fs*width).astype('int')
        mag += np.abs(spec[idx])
    mag /= cnt
    return mag
